src/java/org/apache/cassandra/net/MessageDelivery.java

```diff
@@ -74,9 +85,162 @@ public void onFailure(InetAddressAndPort from, RequestFailureReason reason)
+                                                            Verb verb, REQ request,
+                                                            Iterator<InetAddressAndPort> candidates,
+                                                            TriFunction<Integer, InetAddressAndPort, RequestFailureReason, Boolean> shouldRetry,
+                                                            RetryErrorMessage errorMessage,
+                                                            int attempt)
+    {
+        if (Thread.currentThread().isInterrupted())
+        {
+            promise.tryFailure(new InterruptedException(errorMessage.apply(attempt, ResponseFailureReason.Interrupted, null, null)));
+            return;
+        }
+        if (!candidates.hasNext())
+        {
+            promise.tryFailure(new NoMoreCandidatesException(errorMessage.apply(attempt, ResponseFailureReason.NoMoreCandidates, null, null)));
+            return;
+        }
+        class Request implements RequestCallbackWithFailure<MSG_RSP>
+        {
+            @Override
+            public void onResponse(Message<MSG_RSP> msg)
+            {
+                promise.trySuccess(msgToRsp.apply(attempt, msg));
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailureReason failure)
+            {
+                if (!backoff.mayRetry(attempt))
+                {
+                    promise.tryFailure(new MaxRetriesException(attempt, errorMessage.apply(attempt, ResponseFailureReason.MaxRetries, from, failure)));
+                    return;
+                }
+                if (!shouldRetry.apply(attempt, from, failure))
+                {
+                    promise.tryFailure(new FailedResponseException(from, failure, errorMessage.apply(attempt, ResponseFailureReason.Rejected, from, failure)));
+                    return;
+                }
+                if (promise.isDone() || promise.isCancelled()) return;
+                try
+                {
+                    retryThreads.schedule(() -> sendWithRetries(messaging, promise, msgToRsp, backoff, retryThreads, verb, request, candidates, shouldRetry, errorMessage, attempt + 1),
+                                          backoff.computeWaitTime(attempt), backoff.unit());
+                }
+                catch (Throwable t)
+                {
+                    promise.tryFailure(new FailedScheduleException(errorMessage.apply(attempt, ResponseFailureReason.FailedSchedule, from, failure), t));
+                }
+            }
+        }
+        messaging.sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE), candidates.next(), new Request());
+    }
+
+    enum ResponseFailureReason { MaxRetries, Rejected, NoMoreCandidates, Interrupted, FailedSchedule }
+
+    interface RetryScheduler
+    {
+        void schedule(Runnable command, long delay, TimeUnit unit);
+    }
+
+    enum ImmediateRetryScheduler implements RetryScheduler
```

Comment:
Maybe we should also indicate this would retry on the same thread, WDYT?

Reply:
what do you mean, a name change?  have any in mind?  I went with this name as it matches `Executor`

`org.apache.cassandra.concurrent.ImmediateExecutor`
`io.netty.util.concurrent.ImmediateExecutor`

guava uses the `direct` naming: `com.google.common.util.concurrent.MoreExecutors#directExecutor`

Reply:
you are right; immediate is cool!